File: examples/tick_to_order.py
from tinytrader import Strategy, Contract, NewOrder, OrderStatus, TradeFlag, AutoRun, TradingDay
import statistics
import sys
from simnow import SimnowConfig


class TickToOrder(Strategy):
    def __init__(self, code):
        super().__init__()
        self.instrument = Contract(code)
        print(self.instrument)
        self.latencies = []
        self.sample_count = 300
        self.Subscribe(self.instrument)

    def OnQuote(self, q):
        if q.BidSize1 == 0 or q.AskSize1 == 0:  # 涨跌停不下单
            return

        if len(self.latencies) < self.sample_count:
            newOrder = NewOrder(Instrument=self.instrument, Size=1, Price=q.LowerLimitPrice, Flag=TradeFlag.Open)
            order = self.Insert(newOrder)
            if order is not None:
                # ReceiveTime 为 OnRtnDepthMarketData 第一行的时间戳
                # SendTime 为 ReqOrderInsert 完成后的时间戳
                latency_us = int((order.SendTime - q.ReceiveTime).total_seconds() * 1_000_000)
                self.latencies.append(latency_us)
                print(f"{len(self.latencies):>3d}: {latency_us:>5} us")

    def OnOrder(self, order):
        if order.Status == OrderStatus.Queuing and not order.Canceling:
            self.Cancel(order)
        elif order.Terminated() and len(self.latencies) == self.sample_count:
            self.Report()

    def Report(self):
        sorted_lat = sorted(self.latencies)
        n = len(sorted_lat)
        avg = statistics.mean(sorted_lat)
        min_lat = sorted_lat[0]
        max_lat = sorted_lat[-1]
        p50 = sorted_lat[int(n * 50 / 100)]
        p90 = sorted_lat[int(n * 90 / 100)]
        p95 = sorted_lat[int(n * 95 / 100)]
        p99 = sorted_lat[int(n * 99 / 100)]

        print("\n========== Tick-to-Order 延迟统计 ==========")
        print(f"样本数: {n}")
        print(f"平均值: {avg:.1f} μs")
        print(f"最小值: {min_lat} μs")
        print(f"最大值: {max_lat} μs")
        print(f"P50:    {p50} μs")
        print(f"P90:    {p90} μs")
        print(f"P95:    {p95} μs")
        print(f"P99:    {p99} μs")
        print("============================================")



if __name__ == "__main__":
    config = SimnowConfig()
    config.LogPath = TradingDay() + ".log"
    AutoRun(config, TickToOrder, "rb2701")

File: examples/tick_to_order.cpp
#include "simnow.h"
#include <numeric>

using namespace std;
using namespace chrono;
using namespace tinytrader;

namespace {		// 使用匿名命名空间，防止名称冲突
	class TickToOrder : public Strategy		// 收到行情立即下单，以测试内部延迟
	{
	public:
		TickToOrder(Contract c)
		{
			mInstrument = c;
			Subscribe(mInstrument);
			fmt::print("Subscribe {}\n", c);
		}

		void OnQuote(const Quote& q) override
		{
			if (!q.BidSize1 || !q.AskSize1)		// 涨跌停不下单
				return;

			if (mLatencies.size() < mSampleCount) {
				const Order* order = Insert({
					.Instrument = mInstrument,
					.Size = 1,
					.Price = q.LowerLimitPrice,
					.Flag = TradeFlag::Open
				});
				if (order) {
					// ReceiveTime 为 OnRtnDepthMarketData 第一行的时间戳
					// SendTime 为 ReqOrderInsert 完成后的时间戳
					mLatencies.push_back(order->SendTime - q.ReceiveTime);
					fmt::print("{:>03d}: {:>05.1f} us\n", mLatencies.size(), mLatencies.back().count() / 1000.0);
				}
			}
		}

		void OnOrder(const Order& order) override
		{
			if (order.Status == OrderStatus::Queuing && !order.Canceling)
				Cancel(order);
			else if (order.Terminated() && mLatencies.size() == mSampleCount)
				Report();
		}

	private:
		void Report()
		{
			sort(mLatencies.begin(), mLatencies.end());
			size_t n = mLatencies.size();
			auto to_us = [](nanoseconds ns) -> double {	return ns.count() / 1000.0;	};

			auto total = accumulate(mLatencies.begin(), mLatencies.end(), 0ns);
			double avg = to_us(total) / n;
			double min = to_us(mLatencies.front());
			double max = to_us(mLatencies.back());
			double p50 = to_us(mLatencies[n * 50 / 100]);
			double p90 = to_us(mLatencies[n * 90 / 100]);
			double p95 = to_us(mLatencies[n * 95 / 100]);
			double p99 = to_us(mLatencies[n * 99 / 100]);

			fmt::print("\n========== Tick-to-Order 延迟统计 ==========\n");
			fmt::print("样本数: {}\n", n);
			fmt::print("平均值: {:.1f} μs\n", avg);
			fmt::print("最小值: {:.1f} μs\n", min);
			fmt::print("最大值: {:.1f} μs\n", max);
			fmt::print("P50:    {:.1f} μs\n", p50);
			fmt::print("P90:    {:.1f} μs\n", p90);
			fmt::print("P95:    {:.1f} μs\n", p95);
			fmt::print("P99:    {:.1f} μs\n", p99);
			fmt::print("============================================\n");
		}

		Contract mInstrument;
		vector<nanoseconds> mLatencies;
		size_t mSampleCount = 300;
	};
}


int main()
{
	Config config = SimnowConfig();
	config.LogPath = TradingDay() + ".log";
	return AutoRun<TickToOrder>(config, "rb2701");
}